storage: remove replicaID dependency from sideloaded storage

tbg opened the pull request:
I can't think of a good reason I introduced this initially. Sideloaded
entries belong to Raft entries. Raft entries don't care about the
replicaID, so neither should sideloaded ones. In fact, we had to write
extra code that would move the directories every time the replicaID
changed, and had to reinitialize the sideloaded storage. None of that
should be necessary, which will be a nice simplification once the
migration can be deleted.

Determine the sideloaded directory based on the RangeID alone. The
migration is straightforward: when the new version is active but there's
an old directory, move it to where it ought to be. I factored out the
code that would move the directories prior to the migration to be able
to test that they interact favorably with the migration (not true before
this commit).

NB: the migration can only be deleted safely once we know that there
aren't any deprecated sideloaded directories around when the cluster
version is bumped. But a) this is our expectation and b) it's easy
enough to add a hook to the startup process that enforces it. This
isn't done in this commit, though.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35035)
<!-- Reviewable:end -->

tbg committed:
storage: remove replicaID dependency from sideloaded storage

I can't think of a good reason I introduced this initially. Sideloded
entries belong to Raft entries. Raft entries don't care about the
replicaID, so neither should sideloaded ones. In fact, we had to write
extra code that would move the directories every time the replicaID
changed, and had to reinitialize the sideloaded storage. None of that
should be necessary, which will be a nice simplification once the
migration can be deleted.

Determine the sideloaded directory based on the RangeID alone. The
migration is straightforward: when the new version is active but there's
an old directory, move it to where it ought to be. I factored out the
code that would move the directories prior to the migration to be able
to test that they interact favorably with the migration (not true before
this commit).

NB: the migration can only be deleted safely once we know that there
aren't any deprecated sideloaded directories around when the cluster
version is bumped. But a) this is our expectation and b) it's easy
enough to add a hook to the startup process that enforces it. This
isn't done in this commit, though.

Release note: None

bdarnell commented:
> I can't think of a good reason I introduced this initially.

It was for synchronization purposes (discussed in the RFC PR #16159): What if the replica GC of an old replica races with the creation of a new replica of the same range? Operations on the sideloaded files on disk can't be made atomic with the commit of rocksdb batches, and we wanted to avoid filesystem operations under an important mutex (not sure which one this was referring to)

Maybe there's another way to do this, but that's why it's there now and we shouldn't remove it without understanding and documenting exactly what would make it safe.

tbg commented:
Thanks for digging that up. Honestly I don't think that comment has stood the test of time. The current code does the I/O under all locks

https://github.com/cockroachdb/cockroach/blob/7e0eaf1006c73bd36b1062c8430e807bbfc8049a/pkg/storage/store.go#L2490-L2498

and it's also difficult to imagine it being done without a lock. The Replica owns a part of the keyspace, so we can't let it do I/O while it's potentially sharing that keyspace with some other Replica. We may be able to hold less raftMu on the Replica that is being destroyed, but that doesn't really matter as much because a replica in the process of being destroyed doesn't really have anyone else using the raftMu.

We've also found that a new replica created in place of the old one absolutely needs to see the sideloaded files, and have had to rename the directories accordingly. I think we've just simulated the status quo in a more complicated way.

As an aside, it doesn't seem that the replica GC deletes the sideloaded files at all, oops. We can remedy this now that the truncated state is unreplicated by doing a local log truncation before proceeding to GC.

tbg committed:
logictest: avoid having to edit random files on new cluster versions

Release note: None

bdarnell commented:
> The current code does the I/O under all locks

But not `Store.mu`, which I think might have been my original concern. The fact that we don't unlink the replica on the store until after we've deleted the sideloaded data seems satisfactory to me now.

> As an aside, it doesn't seem that the replica GC deletes the sideloaded files at all, oops.

Are you sure? ReplicaGC calls `RemoveReplica` with `DestroyData: true`, which will call `destroyDataRaftMuLocked` and `postDestroyRaftMuLocked` (which deletes the sideloaded data). We're missing a GC path in the event of a crash between committing the destroy batch and running `postDestroyRaftMuLocked`, but I don't think we're leaking things generally on the replica GC path.

bdarnell reviewed (commented):
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35035)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/replica_sideload_disk.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZ6bbOE-9R5qWkLW62m:-LZ6bbOE-9R5qWkLW62n:bvtm137) ([raw file](https://github.com/cockroachdb/cockroach/blob/657d871dbf213d8c0f954b17d5d4c3ec902e9cfa/pkg/storage/replica_sideload_disk.go#L57)):*
> ```Go
> 		baseDir,
> 		"sideloading",
> 		fmt.Sprintf("r%dXXX", rangeID%1000), // sharding
> ```

Is this unfinished?
___
*[pkg/storage/replica_sideload_disk.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZ6ehMc4t6iiJEDTHtO:-LZ6ehMc4t6iiJEDTHtP:b8tynks) ([raw file](https://github.com/cockroachdb/cockroach/blob/657d871dbf213d8c0f954b17d5d4c3ec902e9cfa/pkg/storage/replica_sideload_disk.go#L73)):*
> ```Go
> }
> 
> func moveSideloadedData(
> ```

This method needs a comment. Are we ever going to be able to get rid of this or does it need to stick around indefinitely for migration purposes?
___
*[pkg/storage/replica_sideload_disk.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZ6exwtDipGbtjz4bAO:-LZ6exwtDipGbtjz4bAP:b-nvy4ln) ([raw file](https://github.com/cockroachdb/cockroach/blob/657d871dbf213d8c0f954b17d5d4c3ec902e9cfa/pkg/storage/replica_sideload_disk.go#L124)):*
> ```Go
> 				return nil, errors.Wrap(err, "creating migrated sideloaded directory")
> 			}
> 			if err := os.Rename(path, newPath); err != nil {
> ```

IIRC this is run for every replica at node startup. Doing 50k stat calls is going to be expensive and all the renames could be even worse. I think we need to arrange this migration in a way that allows us to get rid of the stat calls in the future steady state. 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Are you sure?

No, just missed that, thanks for the pointer.

> But not Store.mu, which I think might have been my original concern.

Oh, ok. Now I see where you're coming from. Yeah, we don't lock that throughout, just at the end to actually unlink the replica. (in the sense, the real lock is having the replica in the store's map because that protects the keyspace from other replicas)

RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-:-LZA1h5REKG-JpQiCGdL:b-pcl1e2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[pkg/storage/replica_sideload_disk.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZ6bbOE-9R5qWkLW62m:-LZ9rl4g2ksOxFVuMBYe:bfw91gm) ([raw file](https://github.com/cockroachdb/cockroach/blob/657d871dbf213d8c0f954b17d5d4c3ec902e9cfa/pkg/storage/replica_sideload_disk.go#L57)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is this unfinished?
</blockquote></details>

No. Added a comment.
___
*[pkg/storage/replica_sideload_disk.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZ6ehMc4t6iiJEDTHtO:-LZ9vqekBnGa90LsrEvR:b-cvajsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/657d871dbf213d8c0f954b17d5d4c3ec902e9cfa/pkg/storage/replica_sideload_disk.go#L73)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This method needs a comment. Are we ever going to be able to get rid of this or does it need to stick around indefinitely for migration purposes?
</blockquote></details>

Comment added. I added migration strategies to the cluster version. TL;DR is we can remove it in 2020.1 or, if we put a little bit of work in, 2019.2.
___
*[pkg/storage/replica_sideload_disk.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZ6exwtDipGbtjz4bAO:-LZ9vz577UOURtfQUv3M:b5xh3o4) ([raw file](https://github.com/cockroachdb/cockroach/blob/657d871dbf213d8c0f954b17d5d4c3ec902e9cfa/pkg/storage/replica_sideload_disk.go#L124)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

IIRC this is run for every replica at node startup. Doing 50k stat calls is going to be expensive and all the renames could be even worse. I think we need to arrange this migration in a way that allows us to get rid of the stat calls in the future steady state. 
</blockquote></details>

Good point, I rearranged things so that we won't `Stat()` when it's already migrated in `moveSideloadedData`, but that's not really going to make a difference (it's not hit at startup). For the call here, I just went ahead and ran a bit of benchmarking (https://github.com/tbg/goplay/tree/master/statbench) that suggests that we don't have to worry about it. I added a comment to that effect as well.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> Reviewed 2 of 3 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35035)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @tbg)

___
*[pkg/storage/replica_sideload_disk.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZ6bbOE-9R5qWkLW62m:-LZBngvY0PKSEGz2XdDt:b-9ko0bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/657d871dbf213d8c0f954b17d5d4c3ec902e9cfa/pkg/storage/replica_sideload_disk.go#L57)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

No. Added a comment.
</blockquote></details>

`1828 % 1000 == 828`. You want division, not mod. 

1000 shards with 64k ranges each only allows for 6.4M range IDs (note that with range merges, we can use up IDs without having that many active ranges at once). This seems too small. If we want to support filesystems with a 64k subdirectory limit, I think we may need three-level sharding. Two-level sharding should have a larger branching factor than 1000. (not a new issue, but while we're changing things and adding documentation that claims to justify the new number we might as well account for our future needs). 
___
*[pkg/storage/replica_sideload_disk.go, line 141 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZBqVtA5KCMwBwaWTCa:-LZBqVtA5KCMwBwaWTCb:b94whcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/280b0dbab1f53ffc728586ac1fc98e9292751592/pkg/storage/replica_sideload_disk.go#L141)):*
> ```Go
> 		// the moment). A stat call for a directory that doesn't exist isn't
> 		// very expensive (on the order of 1000s of ns). For example, on a 2017
> 		// MacBook Pro, this case averages ~3245ns. At 50k replicas, that adds
> ```

Run the benchmark on linux too; that's more important. 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35035)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[pkg/storage/replica_sideload_disk.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZ6bbOE-9R5qWkLW62m:-LZC0CBGEIo1aiPck7-R:buaczw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/657d871dbf213d8c0f954b17d5d4c3ec902e9cfa/pkg/storage/replica_sideload_disk.go#L57)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`1828 % 1000 == 828`. You want division, not mod. 

1000 shards with 64k ranges each only allows for 6.4M range IDs (note that with range merges, we can use up IDs without having that many active ranges at once). This seems too small. If we want to support filesystems with a 64k subdirectory limit, I think we may need three-level sharding. Two-level sharding should have a larger branching factor than 1000. (not a new issue, but while we're changing things and adding documentation that claims to justify the new number we might as well account for our future needs). 
</blockquote></details>

:man_facepalming: thanks for paying attention. I've set the branching factor to 10k but I don't actually think we can get close to running into that limit. The sideloaded storage only has the directory in existence when there are files in the sideloaded storage.

The linux benchmarks (gceworker) came in favorably. Updated the comment to reflect them, but nothing changed in the rationale.
___
*[pkg/storage/replica_sideload_disk.go, line 141 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35035#-LZBqVtA5KCMwBwaWTCa:-LZC2u1A2xeleKEsHxnj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/280b0dbab1f53ffc728586ac1fc98e9292751592/pkg/storage/replica_sideload_disk.go#L141)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Run the benchmark on linux too; that's more important. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=bdarnell

TFTRs!
